File: pi/lib/herdr_workflow/tracing.py
"""Pure traceparent/span math. No I/O, no clock, no network."""
import re
import uuid


def parse_traceparent(value):
    match = re.fullmatch(r"00-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})", value or "", re.I)
    if not match or set(match.group(1)) == {"0"} or set(match.group(2)) == {"0"}:
        return None
    return {"traceId": match.group(1).lower(), "spanId": match.group(2).lower(), "flags": match.group(3).lower()}


def traceparent(context):
    return f"00-{context['traceId']}-{context['spanId']}-{context.get('flags', '01')}"


def child_context(parent=None):
    return {"traceId": parent["traceId"] if parent else uuid.uuid4().hex, "spanId": uuid.uuid4().hex[:16], "flags": parent.get("flags", "01") if parent else "01"}


def span_record(context, name, start_nanos, end_nanos, attributes, parent_span_id=None, status="OK"):
    return {"traceId": context["traceId"], "spanId": context["spanId"], "parentSpanId": parent_span_id, "name": name, "startTimeUnixNano": start_nanos, "endTimeUnixNano": end_nanos, "status": status, "attributes": attributes}

File: agentic-coding/src/workflow/tracing.ts
// Pure traceparent/span math. No I/O, no clock, no network.
import { randomUUID } from 'node:crypto';

export interface TraceContext {
  traceId: string;
  spanId: string;
  flags: string;
}

export interface SpanRecord {
  traceId: string;
  spanId: string;
  parentSpanId: string | null;
  name: string;
  startTimeUnixNano: string;
  endTimeUnixNano: string;
  status: string;
  attributes: Record<string, unknown>;
}

export function parseTraceparent(value: string | null | undefined): TraceContext | null {
  const match = /^00-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$/i.exec(value || '');
  if (!match) return null;
  const [, traceId, spanId, flags] = match;
  if (/^0+$/.test(traceId) || /^0+$/.test(spanId)) return null;
  return { traceId: traceId.toLowerCase(), spanId: spanId.toLowerCase(), flags: flags.toLowerCase() };
}

export function traceparent(context: TraceContext): string {
  return `00-${context.traceId}-${context.spanId}-${context.flags ?? '01'}`;
}

export function childContext(parent?: TraceContext | null): TraceContext {
  return {
    traceId: parent ? parent.traceId : randomUUID().replace(/-/g, ''),
    spanId: randomUUID().replace(/-/g, '').slice(0, 16),
    flags: parent ? parent.flags ?? '01' : '01',
  };
}

export function spanRecord(
  context: TraceContext,
  name: string,
  startNanos: string,
  endNanos: string,
  attributes: Record<string, unknown>,
  parentSpanId: string | null = null,
  status = 'OK',
): SpanRecord {
  return {
    traceId: context.traceId,
    spanId: context.spanId,
    parentSpanId,
    name,
    startTimeUnixNano: startNanos,
    endTimeUnixNano: endNanos,
    status,
    attributes,
  };
}
